couler/core/proto_repr.py

```diff
@@ -132,16 +132,18 @@ def step_repr(
             if isinstance(arg, OutputArtifact):
                 pb_art = pb_step.args.add()
                 pb_art.artifact.name = arg.artifact["name"]
-                pb_art.artifact.value = (
-                    '"{{inputs.artifacts.%s}}"' % pb_art.artifact.name
-                )
+                pb_art.artifact.value = arg.value
             else:
                 pb_param = pb_step.args.add()
                 pb_param.parameter.name = utils.input_parameter_name(
                     pb_step.name, i
                 )
                 pb_param.parameter.value = (
-                    '"{{inputs.parameters.%s}}"' % pb_param.parameter.name
+                    # This is casted to string for protobuf and Argo
+                    # will automatically convert it to a correct type.
+                    str(arg)
```

Comment:
but how about other engines?

Reply:
We can infer and cast it to the correct type during the conversion from IR to engine specific presentation.